sqldb/interfaces.go

```diff
@@ -0,0 +1,258 @@
+	ReadOnly() bool
+}
+
+// BatchedTx is a generic interface that represents the ability to execute
+// several operations to a given storage interface in a single atomic
+// transaction. Typically, Q here will be some subset of the main sqlc.Querier
+// interface allowing it to only depend on the routines it needs to implement
+// any additional business logic.
+type BatchedTx[Q any] interface {
+	// ExecTx will execute the passed txBody, operating upon generic
+	// parameter Q (usually a storage interface) in a single transaction.
+	//
+	// The set of TxOptions are passed in order to allow the caller to
+	// specify if a transaction should be read-only and optionally what
+	// type of concurrency control should be used.
+	ExecTx(ctx context.Context, txOptions TxOptions,
+		txBody func(Q) error) error
+}
+
+// Tx represents a database transaction that can be committed or rolled back.
+type Tx interface {
+	// Commit commits the database transaction, an error should be returned
+	// if the commit isn't possible.
+	Commit() error
+
+	// Rollback rolls back an incomplete database transaction.
+	// Transactions that were able to be committed can still call this as a
+	// noop.
+	Rollback() error
+}
+
+// QueryCreator is a generic function that's used to create a Querier, which is
+// a type of interface that implements storage related methods from a database
+// transaction. This will be used to instantiate an object callers can use to
+// apply multiple modifications to an object interface in a single atomic
+// transaction.
+type QueryCreator[Q any] func(*sql.Tx) Q
+
+// BatchedQuerier is a generic interface that allows callers to create a new
+// database transaction based on an abstract type that implements the TxOptions
+// interface.
+type BatchedQuerier interface {
+	// Querier is the underlying query source, this is in place so we can
+	// pass a BatchedQuerier implementation directly into objects that
+	// create a batched version of the normal methods they need.
+	sqlc.Querier
+
+	// BeginTx creates a new database transaction given the set of
+	// transaction options.
+	BeginTx(ctx context.Context, options TxOptions) (*sql.Tx, error)
+}
+
+// txExecutorOptions is a struct that holds the options for the transaction
+// executor. This can be used to do things like retry a transaction due to an
+// error a certain amount of times.
+type txExecutorOptions struct {
+	numRetries int
+	retryDelay time.Duration
+}
+
+// defaultTxExecutorOptions returns the default options for the transaction
+// executor.
+func defaultTxExecutorOptions() *txExecutorOptions {
+	return &txExecutorOptions{
+		numRetries: DefaultNumTxRetries,
+		retryDelay: DefaultRetryDelay,
+	}
+}
+
+// randRetryDelay returns a random retry delay between 0 and the configured max
+// delay.
+func (t *txExecutorOptions) randRetryDelay() time.Duration {
+	return time.Duration(prand.Int63n(int64(t.retryDelay))) //nolint:gosec
```

Comment:
One thing we want to fix in the repo this originated was to make this actually have a min back off, as is right now, it can use a rather small value (say 1 ns) which causes the retries to bleed out pretty quickly. 

So I think we want to also have a min value here, then offset from that, 50 ms for a min value seems good I think? Then we can add a TODO here to circle back on doing proper exponential backoff as well.

Reply:
Or perhaps we can also try and do an exponential ramp up?